bindinfo/bind_cache.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bindinfo
+
+import (
+	"sync"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/kvcache"
+	"github.com/pingcap/tidb/util/memory"
+)
+
+// bindCache uses the LRU cache to store the bindRecord.
+// The key of the LRU cache is original sql, the value is a slice of BindRecord.
+// Note: The bindCache is thread-safe.
+type bindCache struct {
+	lock        sync.Mutex
+	cache       *kvcache.SimpleLRUCache
+	memCapacity int64
+	memTracker  *memory.Tracker // track memory usage.
+}
+
+type bindCacheKey string
+
+func (key bindCacheKey) Hash() []byte {
+	return hack.Slice(string(key))
+}
+
+func bindCacheKVMem(key bindCacheKey, value []*BindRecord) int64 {
+	var valMem int64
+	for _, bindRecord := range value {
+		valMem += int64(bindRecord.size())
+	}
+	return int64(len(key.Hash())) + valMem
+}
+
+func newBindCache(memCapacity int64) *bindCache {
+	// since bindCache controls the memory usage by itself, set the capacity of
+	// the underlying LRUCache to max to close its memory control
+	cache := kvcache.NewSimpleLRUCache(mathutil.MaxUint, 0.1, 0)
+	c := bindCache{
+		cache:       cache,
+		memCapacity: memCapacity,
+		memTracker:  memory.NewTracker(memory.LabelForBindCache, -1),
+	}
+	return &c
+}
+
+// get gets a cache item according to cache key. It's not thread-safe.
+// Note: Only other functions of the bindCache file can use this function.
+// Don't use this function directly in other files in bindinfo package.
+// The return value is not read-only, but it is only can be used in other functions which are also in the bind_cache.go.
+func (c *bindCache) get(key bindCacheKey) []*BindRecord {
+	value, hit := c.cache.Get(key)
+	if !hit {
+		return nil
+	}
+	typedValue := value.([]*BindRecord)
+	return typedValue
+}
+
+// set inserts an item to the cache. It's not thread-safe.
+// Only other functions of the bindCache can use this function.
+func (c *bindCache) set(key bindCacheKey, value []*BindRecord) bool {
```

Comment:
For the behavior when the cache exceeds the memory capacity, do we have a design?